### PR TITLE
Use a more accurate range for No matching overload

### PR DESCRIPTION
**Pull request by edgarfgp**
Attempts to Fix https://github.com/dotnet/fsharp/issues/14284 as part of the [Amplifying F#](https://amplifying-fsharp.github.io/sessions/2023/03/03/)  initiative. [Live session](https://www.youtube.com/watch?v=rL5RrJUiBvE&t=926s)

### CODE REVIEW

File: src/Compiler/Service/ServiceDeclarationLists.fs [hunk 1]
```diff
@@ -337,7 +337,7 @@ module DeclarationListHelpers =
 
         // F# constructors and methods
         | Item.CtorGroup(_, minfos) 
-        | Item.MethodGroup(_, minfos, _) ->
+        | Item.MethodGroup(_, _, minfos, _) ->
```
Comment (nojaf) [inline src/Compiler/Service/ServiceDeclarationLists.fs:340]: Nitpick: named access, please 😇

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs [hunk 2]
```diff
@@ -686,4 +686,27 @@ type X =
...
+    static member Instance = T()
+
+    member _.Method(_: double) = ()
+    member _.Method(_: int) = ()
+
+T.Instance.Method("")
+        """
+        |> typecheck
+        |> shouldFail
+        |> withDiagnostics [
```
Comment (nojaf) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:704]: I think you can use the `|> withSingleDiagnostic` here.

File: src/Compiler/Checking/NameResolution.fs [hunk 3]
```diff
@@ -269,8 +269,8 @@ type Item =
         | Item.Event einfo -> einfo.DisplayNameCore 
         | Item.Property(_, pinfo :: _) -> pinfo.DisplayNameCore
         | Item.Property(nm, _) -> nm |> ConvertValLogicalNameToDisplayNameCore
-        | Item.MethodGroup(_, FSMeth(_, _, v, _) :: _, _) -> v.DisplayNameCore
-        | Item.MethodGroup(nm, _, _) -> nm |> ConvertValLogicalNameToDisplayNameCore
+        | Item.MethodGroup(methods = FSMeth(_, _, v, _) :: _) -> v.DisplayNameCore
+        | Item.MethodGroup(displayName= nm) -> nm |> ConvertValLogicalNameToDisplayNameCore
```
Comment (nojaf) [inline src/Compiler/Checking/NameResolution.fs:273]: Another nitpick but named access is formatted as `X(y = z)`.
See https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-named-arguments

File: src/Compiler/Checking/CheckExpressions.fs [hunk 4]
```diff
@@ -9201,6 +9201,7 @@ and TcMethodApplicationThen
        m           // The range of the object argument or whole application. We immediately union this with the range of the arguments
        mItem       // The range of the item that resolved to the method name
        methodName // string, name of the method
+       (methodNameRange: range option)
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckExpressions.fs:9204]: Why not use/fix `mItem` range?

**Review by auduchinok — CHANGES_REQUESTED** [review]
@edgarfgp Have you considered fixing existing `mItem` range instead of adding another one?

This solutions seems to be a bit too-adhoc for this problem, and it only seems to hide/workaround the underlying problem instead of fixing it, moreover it added an excessive field that already partially duplicates existing data. Fixing it here could be a good start of having more name resolution range fixes later, that would also include qualified identifiers, type applications, function/method applications and so on.

File: src/Compiler/Checking/CheckExpressions.fs (hunk 4 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs:9204]: Thanks for the review. I Agree fixing `mItem` range seems a proper fix. Will have a look

Comment (nojaf) [reply src/Compiler/Checking/CheckExpressions.fs:9204]: The original reason (at least from my end) was to add a new range as it was a bit of a guess on what would happen when we update the existing `m`. I couldn't really asses the impact so it felt safer to just start with a very specific fix and then expand on that.

**Pull request closed without merging**